### DISCUSSION TITLE
Visualize a Custom Object - J2000 list of positions and time

### DISCUSSION THREAD

**Discussion by sgano** [Stellarium/stellarium#2203] [General]
Hello - I am trying to find a way to visualize a custom object based on a ephemeris table.  After searching through the documentation have not been able to figure it out.   Specifically I have a table of J2000.0 [time, X, Y, Z] values for an object and would like to be able to see what this would look like on different points on Earth and at different times (within the range provided).  Is there a way to create an object given such a table?  Or through a script can I create an object that moves to a new position (which I can interpolate) based on the time the user changes the scene to?    I would appreciate any pointers to documentation or ideas.   

Alternately I could do this manually for a few specific locations and times, but I also haven't been able to find a way to enter a J2000 position using X,Y,Z coordinates directly (is that possible?). 

Thanks!

**Comment by stephanpls** [comment]
I doubt if you can plot XYZ coordinates in Stellarium. Is your custom object a satellite, an asteroid, a comet etc.? Then you could check at https://ssd.jpl.nasa.gov/horizons/ if it provides Kepler orbital elements for your object. Then put these in a Stellarium .ini file and off you go.
No orbital elements provided? Even then it's possible but it means quite some celestial math. The trick is to convert your XYZ to right ascension and declination, but you need the Sun's XYZ as well. Jean Meeus's 'Astronomical Algorithms' is a very useful book for this. 
Having both the XYZ of your object and the Sun, calculate the right ascension and declination, and put them in a script using statement core.moveToRaDecJ2000(RA, DC). Quite some work to do!

**Reply by sgano** [reply]
Thanks for the reply and glad to hear confirmation that this was not directly possible.  It is for an upcoming satellite mission so no orbital elements on the usual sites.  I ended up doing something similar to your recommendation and wrote up some python scripts to calculate the local right ascension and declination for different locations on Earth -- then plotted that in Stellarium.

**Reply by stephanpls** [reply]
Nice to hear you succeeded. Quite interested what type of satellite mission... For telecom? For internet to the remotest places? Sorry if I ask too much.

**Reply by sgano** [reply]
Specifically I was looking at the Artemis I mission.   I was trying to determine the position, relative to the moon, the Orion spacecraft would appear on it's outbound journey to the moon -- and how that varied from different places around the Earth.

**Reply by Rubiksmath** [reply]
Funny this was for Artemis I because I actually had exactly the same question for Artemis II. I spent a while making some python scripts to try and estimate the local Az-Alt of the spacecraft. This was done from very limited information, with a few outdated and crude estimates of orbital apoapsis and periapsis, time of burns, etc, and with code that is very cobbled together and rushed with no structure whatsoever. It turned out not that bad honestly, when the Horizons ephemeris came online some time later, (unless it was there before and I missed it), I compared them and it was pretty decent. As in, off by quite a bit, but my predictions still show strong resemblance on a polar plot (and they give predictions starting at T+08m18s, whereas the Horizons only comes online after a few hours).

When I found that the Horizons ephemeris existed, I was very excited because I thought I could import it into Stellarium. However I cannot find a way to do this, which is really frustrating given that the ephemeris outright give you the local (or geocentric) RA and DEC for free, all that Stellarium would have to do (in my view) would be to place an object there.

It would be *really* nice if there is actually some way to do this, or if there isn't that this capability pretty please be added? Maybe not the best place to "ask" but I just came across this and thought I had to comment.

**Reply by 10110111** [reply]
@Rubiksmath I think you want to read the User Guide chapter D.2.3 JPL Horizons.

**Reply by gzotti** [reply]
We do currently not support anything that cannot be expressed in terms of Kepler elements around the Sun or a planet, or as TLE set for satellites in stable (mostly unpowered) orbits. If you have a system that computes positions, maybe you can write something that sets Markers via our RemoteControl API. If there is an application for it, a "PoweredSolarSystemObject" could be developed that could be placed by scripting commands, so could be controlled by your external software. (It should include a numerical integrator that applies nearby-object gravity.) I don't know about live data sources that emit state vectors (time, position, speed) for ongoing space missions, though.  Volunteers welcome...

**Reply by sgano** [reply]
I have come up with a script that takes an OEM ephemeris file for Artemis II, and generates the data needed for a Stellarium script (.ssc and .inc) I created which basically monitors for time changes and then updates the position of a custom object.    That script can be found here:   https://www.gano.name/shawn/artemis2/     I will try to make it more generic in the future.

**Reply by sgano** [reply]
I also describe how to use it in this video:  https://www.youtube.com/watch?v=GqiRlgBeyWY
